### PR TITLE
feat: ini rust repo & `wallet-metamask` crate

### PR DESCRIPTION
**Pull request by shunkakinoki**
initial push for initializing `wallet-rs`

this pr does the following:

- [x] migrate from mm's [browser-password](https://github.com/MetaMask/browser-passworder/blob/a8574c40d1e42b2bc2c2b3d330b0ea50aa450017/src/index.ts)
- [x] migrate from mm's [vault-decrypt](https://github.com/MetaMask/vault-decryptor)
- [x] initialize repo structure and crates
- [x] added tests and specs
- [x] added initial devops

### CODE REVIEW

File: .cargo/config.toml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+[alias]
+uniffi-bindgen = ["run", "--package", "embedded-uniffi-bindgen", "--"]
+mpl-license-checker = ["run", "--package", "mpl-license-checker", "--"]
+
+[target.aarch64-apple-darwin]
+rustflags = ["-C", "link-arg=-fuse-ld=/opt/homebrew/opt/llvm/bin/ld64.lld"]
```
Comment (stanleyjzheng) [inline .cargo/config.toml]: Super NIT: Should this be in the repo's `config.toml`? Perhaps would be more suited in `~/.cargo/config.toml` for those who use other linkers

File: crates/metamask/src/password.rs [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+    let mut rng = OsRng;
+    let mut bytes = [0u8; 16];
+    rng.fill_bytes(&mut bytes);
+    let nonce = Nonce::from_slice(&bytes);
+
+    // Construct a key from the provided bytes.
+    let cipher = Aes256Gcm::new(key.into());
+
+    // Encrypt the data.
+    let data = cipher.encrypt(nonce, data.as_ref()).unwrap();
```
Comment (stanleyjzheng) [inline crates/metamask/src/password.rs]: Not a fan of the unwrap unless it's unreachable -- maybe we can do something along the lines of `let data = cipher.encrypt(nonce, data.as_ref()).map_err(|e| e.to_string())?;`?

File: crates/metamask/src/password.rs [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+
+    // Return the encrypted data.
+    let text = Vault {
+        data: general_purpose::STANDARD.encode(data),
+        iv: general_purpose::STANDARD.encode(nonce),
+        salt,
+    };
+
+    Ok(serde_json::to_string(&text).unwrap())
```
Comment (stanleyjzheng) [inline crates/metamask/src/password.rs:62]: Ditto, maybe `Ok(serde_json::to_string(&text)?)` (and similarly to other places)

Feel free to ignore if it's safe

**Review by stanleyjzheng — APPROVED** [review]
Looks great, super exciting! I have a couple more general suggestions

Fist, I would highly suggest `anyhow::Result<T>` over `Result<T, Box<dyn Error>>`, see https://antoinerr.github.io/blog-website/2023/01/28/rust-anyhow.html or https://users.rust-lang.org/t/why-i-use-anyhow-error-even-in-libraries/68592 etc. 

Also, maybe it's worth considering more strict unwrap/expect linting, like

```toml
[build]
rustflags = [
  "-Dwarnings",
  "-Dclippy::expect_used",
  "-Dclippy::unwrap_used",
]
```

in `.cargo/config.toml`. Then adding 

```rs
#![allow(clippy::expect_used)]
#![allow(clippy::unwrap_used)]
```

on test modules, and `#[allow(clippy::unwrap_used)]` in front of known safe unwraps.

File: crates/metamask/src/password.rs (hunk 3 above)
Comment (shunkakinoki) [reply crates/metamask/src/password.rs:62]: Ahh omg thank you rust saviour 😭😭😭

File: crates/metamask/src/password.rs (hunk 2 above)
Comment (shunkakinoki) [reply crates/metamask/src/password.rs]: Ahhhhhhhh this makes so much sense. Thank you so much. Jesus.

File: .cargo/config.toml (hunk 1 above)
Comment (shunkakinoki) [reply .cargo/config.toml]: Ahh true. You're right. Removing.

**Pull request merged**